p2p/libp2p/identity/identity.go

```diff
@@ -0,0 +1,125 @@
+package identity
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multiaddr"
+
+	"github.com/pokt-network/pocket/p2p/common"
+	"github.com/pokt-network/pocket/p2p/transport"
+	"github.com/pokt-network/pocket/p2p/types"
+	poktCrypto "github.com/pokt-network/pocket/shared/crypto"
+)
+
+var (
+	ErrIdentity = common.NewErrFactory("")
+)
+
+// PoktPeerFromStream builds a pokt peer from a libp2p stream.
+// (NOTE: excludes `ServiceURL`)
+func PoktPeerFromStream(stream network.Stream) (*types.NetworkPeer, error) {
+	remotePubKeyBytes, err := stream.Conn().RemotePublicKey().Raw()
+	// NB: abort handling this stream.
+	if err != nil {
+		return nil, err
+	}
+	poktPubKey, err := poktCrypto.NewPublicKeyFromBytes(remotePubKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.NetworkPeer{
+		Dialer:    transport.NewLibP2PTransport(stream),
+		PublicKey: poktPubKey,
+		// NB: pokt analogue of libp2p peer.ID
+		Address:    poktPubKey.Address(),
+		ServiceUrl: stream.Conn().RemoteMultiaddr().String(),
```

Comment:
Does this return a problem url?

Reply:
`We could of course convert the multiaddr to a proper URL here, if you think that would be a better approach.`

I think this makes more sense